### PR TITLE
Gear vendor UI

### PR DESCRIPTION
**Pull request by mullenpaul**
## About The Pull Request

Migrates gear vendors to TGUI. Massive refactor to how cm_vendors work, unifying the code and making all operate in a standard way. Vendors now describe the checks and actions they want carried out.

## Why It's Good For The Game

Vendor TGUI part 3, finishing all cm_vendors. Massive code refactoring making things a lot more unified.

## Changelog

:cl:
refactor: gear vendors now use tgui
refactor: complete rewrite of how vendors operate
/:cl:

### CODE REVIEW

File: code/game/machinery/vending/vendor_types/food.dm [hunk 1]
```diff
@@ -9,20 +9,12 @@
...
-
-			list("FLASKS", -1, null, null),
-			list("Metal Flask", 10, /obj/item/reagent_container/food/drinks/flask, VENDOR_ITEM_REGULAR),
-			list("USCM Flask", 5, /obj/item/reagent_container/food/drinks/flask/marine, VENDOR_ITEM_REGULAR)
-			)
+/obj/structure/machinery/cm_vending/sorted/marine_food/get_listed_products(mob/user)
+	. = list()
+	. += GLOB.cm_vending_marine_food
+	//. += GLOB.cm_vending_marine_food_christmas
+	. += GLOB.cm_vending_marine_food_flasks
```
Comment (fira) [inline code/game/machinery/vending/vendor_types/food.dm:16]: mind this creates a new list so should be avoided for bigger vendors

should integrate with the xmas food decorator somehow if possible

**Review by ThePiachu — CHANGES_REQUESTED** [review]
The vehicle parts vendor's point costs get a bit cut off:

image

File: code/game/machinery/vending/cm_vending.dm [hunk 2]
```diff
@@ -251,7 +268,158 @@ GLOBAL_LIST_EMPTY(vending_products)
...
+
+/obj/structure/machinery/cm_vending/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+
+	var/mob/living/carbon/human/user = usr
+	switch (action)
+		if ("cancel")
+			SStgui.close_uis(src)
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/game/machinery/vending/cm_vending.dm:286]: I think you're right

File: code/game/machinery/vending/cm_vending.dm [hunk 3]
```diff
@@ -251,7 +268,158 @@ GLOBAL_LIST_EMPTY(vending_products)
...
+								to_chat(user, SPAN_WARNING("Only USCM Synthetics may vend experimental tool tokens."))
+								vend_fail()
+								return FALSE
+
+					if(!handle_vend(itemspec, user))
+						to_chat(user, SPAN_WARNING("You can't buy things from this category anymore."))
+						vend_fail()
+						return FALSE
+
+			if(use_points|use_snowflake_points)
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/game/machinery/vending/cm_vending.dm:371]: Good catch, thanks!

File: code/game/machinery/vending/vendor_types/food.dm (hunk 1 above)
Comment (mullenpaul) [reply code/game/machinery/vending/vendor_types/food.dm:16]: I've reworked this so the food is added to the populate_product_list proc - this is only ran on init, so the decorator will override the listed_products.

**Comment by mullenpaul** [comment]
Quick resolution was to do this https://media.discordapp.net/attachments/745447048261795890/1045796921475874926/image.png?width=306&height=611

**Review by morrowwolf — CHANGES_REQUESTED** [review]
RTO headsets not vending correctly. May be a wider issue.

**Review by ThePiachu — CHANGES_REQUESTED** [review]
The issues I heard about this round:
- Various Squad roles didn't get the correct headset, I think all of them just got the detault one
- APC reportedly didn't have the frontal cannons in the vendor, but I checked it locally and it worked fine, so not sure what's up
- Comtechs got regular marine gloves rather than some special gloves they were supposed to?
- "compared to the marine equipment racks, the squad vendors no longer close their windows when you walk away from them anymore"

**Comment by ThePiachu** [comment]
Also, I think the standard procedure is to Rebase a PR, not merge Master in, so for future reference please don't do that.

**Comment by mullenpaul** [comment]
Thanks for the feedback, I think the issues could be because I moved the local item list to a global one - the vendors maybe didn't do the change. I'll look into and resolve the issue asap.

If I rebase a pushed branch isn't it going to break the history of it? I've had issues in the past and the pr history was all over the place

**Comment by mullenpaul** [comment]
I have resolved the issues with the headsets/gloves

**Comment by mullenpaul** [comment]
I have resolved the issues with the headsets/gloves

**Comment by mullenpaul** [comment]
I've resolved the issues regarding vendors not closing. I allowed the req vendor windows to stay open - based off feedback from req players. The subclasses needed a custom ui state.

The APC issues can't be replicated.

**Review by ThePiachu — CHANGES_REQUESTED** [review]
There seems to be issue with vending prepared meals in the food vendor, and I quote - "oh god oh fuck, what have you done to the food vendors, they run out as soon as someone vends something.".

They also seem to go into negatives:
image

Doesn't seem to affect the Flasks though...

**Comment by mullenpaul** [comment]
The food vendor issues have been resolved - the sorted vendor was referencing a global list for its inventory and the quantity available was ship wide.

**Comment by mullenpaul** [comment]
apc issues have been fixed as well as issue reported with CO weapon

File: code/game/machinery/vending/vendor_types/requisitions.dm [hunk 4]
```diff
@@ -667,12 +662,18 @@ obj/structure/machinery/cm_vending/sorted/uniform_supply
...
+
 /obj/structure/machinery/cm_vending/sorted/uniform_supply/populate_product_list(var/scale)
 	return
 
 /obj/structure/machinery/cm_vending/sorted/uniform_supply/vend_fail()
 	return
 
+/obj/structure/machinery/cm_vending/sorted/uniform_supply/ui_state(mob/user)
+	return GLOB.not_incapacitated_and_adjacent_state
+
```
Comment (stanalbatross) [inline code/game/machinery/vending/vendor_types/requisitions.dm:676]: duped proc that should be inherited? I'm seeing a few /ui_state procs that look extraneous

File: code/game/machinery/vending/vendor_types/squad_prep/squad_prep.dm [hunk 5]
```diff
@@ -8,6 +8,9 @@
 	req_one_access = list(ACCESS_MARINE_LOGISTICS, ACCESS_MARINE_PREP, ACCESS_MARINE_CARGO)
 	hackable = TRUE
 
+/obj/structure/machinery/cm_vending/sorted/cargo_guns/squad_prep/ui_state(mob/user)
+	return GLOB.not_incapacitated_and_adjacent_strict_state
```
Comment (stanalbatross) [inline code/game/machinery/vending/vendor_types/squad_prep/squad_prep.dm:12]: same here

**Review by stanalbatross — CHANGES_REQUESTED** [review]

**Review by ThePiachu — CHANGES_REQUESTED** [review]
This PR went in a bit early, please check if the changes work with your PR or does it need to be re-applied - https://github.com/cmss13-devs/cmss13/pull/1720 .

**Comment by ThePiachu** [comment]
Alright, I think the PR will be good to merge once the outstanding issues are resolved!

File: code/game/machinery/vending/vendor_types/requisitions.dm (hunk 4 above)
Comment (mullenpaul) [reply code/game/machinery/vending/vendor_types/requisitions.dm:676]: It's good in this case. Removing the suplicate definition here.

File: code/game/machinery/vending/vendor_types/squad_prep/squad_prep.dm (hunk 5 above)
Comment (mullenpaul) [reply code/game/machinery/vending/vendor_types/squad_prep/squad_prep.dm:12]: As discussed on discord - this is to allow for req vendors to be non-strict but squad prep vendors to be strict

**Review by harryob — APPROVED** [review]

**Pull request merged**